README.md
# DNAfusion <img src="vignettes/logo.png" width="150" align="right">
[![platform](http://www.bioconductor.org/shields/availability/devel/DNAfusion.svg)](https://bioconductor.org/packages/release/bioc/html/DNAfusion.html#archives)
[![](https://img.shields.io/badge/release%20version-3.4.2-green.svg)](https://bioconductor.org/packages/release/bioc/html/DNAfusion.html)
[![Anaconda-Server Badge](
https://anaconda.org/bioconda/bioconductor-dnafusion/badges/downloads.svg)](
https://anaconda.org/bioconda/bioconductor-dnafusion/badges/downloads.svg)
[![Hits](https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2FCTrierMaansson%2FDNAfusion&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false)](https://hits.seeyoufarm.com)

An R/Bioconductor package to evaluate BAM files and identify genefusions, such as EML4-ALK

> This package was created in order to increase the sensitivity of EML4-ALK detection from commercially available NGS products such as the AVENIO (Roche) pipeline. Paired-end sequencing of cfDNA generated BAM files can be used as input to discover EML4-ALK variants. This package was developed using position deduplicated BAM files generated with the AVENIO Oncology Analysis Software. These files are made using the AVENIO ctDNA surveillance kit and Illumina Nextseq 500 sequencing. This is a targeted hybridization NGS approach and includes ALK-specific but not EML4-specific probes. The package includes six functions. The output of the first function, `EML4_ALK_detection()`, is used to determine whether EML4-ALK is detected and serves as input for the next four  exploratory functions characterizing the EML4-ALK variant. The last function `EML4_ALK_analysis()` combines the output of the exploratory functions. To serve as examples, this package includes BAM files representing the EML4-ALK positive cell line H3122 and the EML4-ALK negative cell line, HCC827.

## Highlights
DNAfusion is under active development. In the future the package will include more genefusions such as ROS1 and RET, as well as other ALK fusionpartners.

## Installation

Use devtools to install the most recent version of DNAfusion from the GitHub repository.

```R
if (!require(devtools)) install.packages('devtools')
library(devtools)

devtools::install_github("CTrierMaansson/DNAfusion", build_vignettes = TRUE)
library(DNAfusion)

```

Alternatively, install DNAfusion published at Bioconductor.

```R
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DNAfusion")
library(DNAfusion)

```

## Vignettes

See the documentation at https://github.com/CTrierMaansson/DNAfusion/tree/main/vignettes

Alternatively, the vignettes can be browsed in Rstudio with 

 ```R
browseVignettes("DNAfusion")
```
This requires `build_vignettes = TRUE` during installation with github

## Citation

If you use [DNAfusion](https://bioconductor.org/packages/release/bioc/html/DNAfusion.html) 
in published research please cite:

1.  Christoffer Trier Maansson, Emma Roger Andersen, Maiken Parm Ulhoi, Peter Meldgaard, Boe Sandahl Sorensen (2022). DNAfusion: An R/Bioconductor package to identify gene fusions. R package version 1.2.0 <https://github.com/CTrierMaansson/DNAfusion>

2.  Maansson CT, Andersen ER, Ulhoi MP, Meldgaard P, Sorensen BS. “DNAfusion: an
R/Bioconductor package for increased sensitivity of detecting gene fusions in liquid
biopsies.” BMC Bioinformatics, 2023, 24:131,
[doi:10.1186/s12859-023-05259-3](https://doi.org/10.1186/s12859-023-05259-3).


NEWS.md
# DNAfusion 1.3.0 (5 April 2023)
* Added introns_ALK_EML4()
* Added find_variants()
* Updated break_position_depth()
* Updated break_position()
* Refined EML4_ALK_detection()
* Updated vignette to include new functions
* Cleaned functions for Bioconductor 3.17 release

# DNAfusion 1.1.1 (6 March 2023)
* Updated example files to include ALK reads

# DNAfusion 1.1.0 (13 December 2022)
* Updated DESCRIPTION

# DNAfusion 1.0.0 (19 October 2022)
* Bioconductor 3.16 release 

# DNAfusion 0.99.3 (16 October 2022)
* Updated DESCRIPTION

# DNAfusion 0.99.2 (14 Octoboer 2022)
* Review from Bioconductor

# DNAfusion 0.99.1 (12 September 2022)
* Changed the output of EML4_ALK_detection() from data.frame to GAlignments
object
* Adjusted the other functions, vignettes, tests and man accordingly

# DNAfusion 0.99.0 (31 August 2022)
* DNAfusion takes BAM file as input and identifies EML4-ALK gene fusions
* Submitted to Bioconductor.


tests/testthat.R
library(DNAfusion)
library(testthat)
test_check("DNAfusion")


tests/testthat/test_functions.R
testthat::context("Test of DNAfusion functions")
library(DNAfusion)

file <- system.file("extdata","HCC827_EML4.bam", package = "DNAfusion")
gene <-  "EML4"
detection <- EML4_ALK_detection(file = file)
EML4_seq <- EML4_sequence(EML4_ALK_detection(file = file))
ALK_seq <- ALK_sequence(EML4_ALK_detection(file = file))
pos <- break_position(EML4_ALK_detection(file = file), gene=gene)
pos_depth <- break_position_depth(file = file, EML4_ALK_detection(file = file),
                                  gene=gene)
analysis <- EML4_ALK_analysis(file = file)

testthat::expect_true(is.character(file))
testthat::expect_true(is.character(gene))
testthat::expect_true(isa(detection, "GAlignments"))
testthat::expect_true(is.table(EML4_seq) | is.character(EML4_seq))
testthat::expect_true(is.table(ALK_seq) | is.character(ALK_seq))
testthat::expect_true(is.table(pos) | is.character(pos))
testthat::expect_true(is.integer(pos_depth) | is.character(pos_depth))
testthat::expect_true(is.list(analysis) | isa(analysis, "GAlignments"))


file <- system.file("extdata","H3122_EML4.bam", package = "DNAfusion")
gene <-  "EML4"
detection <- EML4_ALK_detection(file = file)
EML4_seq <- EML4_sequence(EML4_ALK_detection(file = file))
ALK_seq <- ALK_sequence(EML4_ALK_detection(file = file))
pos <- break_position(EML4_ALK_detection(file = file), gene=gene)
pos_depth <- break_position_depth(file = file, EML4_ALK_detection(file = file),
                                  gene=gene)
analysis <- EML4_ALK_analysis(file = file)

testthat::expect_true(is.character(file))
testthat::expect_true(is.character(gene))
testthat::expect_true(isa(detection, "GAlignments"))
testthat::expect_true(is.table(EML4_seq) | is.character(EML4_seq))
testthat::expect_true(is.table(ALK_seq) | is.character(ALK_seq))
testthat::expect_true(is.table(pos) | is.character(pos))
testthat::expect_true(is.integer(pos_depth) | is.character(pos_depth))
testthat::expect_true(is.list(analysis) | isa(analysis, "GAlignments"))
testthat::expect_error(break_position(detection,gene = "character"))
testthat::expect_error(break_position(detection,gene = 1))
testthat::expect_error(break_position(detection,genome = "38",gene=1))
testthat::expect_error(break_position(detection,genome = "38",gene="character"))
testthat::expect_error(break_position(detection,genome = 38,gene=1))
testthat::expect_error(break_position(detection,genome = 38,gene="character"))
testthat::expect_error(break_position_depth(file=file,detection,gene=1))
testthat::expect_error(break_position_depth(file=file, detection,
                                            gene="character"))
testthat::expect_error(break_position_depth(file=file, detection,genome = 38))
testthat::expect_error(break_position_depth(file=file, detection,genome = "38"))
testthat::expect_error(ALK_sequence(detection, basepairs = "20"))
testthat::expect_error(EML4_sequence(detection, basepairs = "20"))

testthat::expect_error(EML4_ALK_detection(file = file, genome = 38))
testthat::expect_error(EML4_ALK_detection(file = file, mates = "38"))
testthat::expect_error(EML4_ALK_detection(file = file, genome = "hg37"))


testthat::expect_error(ALK_sequence(detection,genome="38"))
testthat::expect_error(ALK_sequence(detection,genome=38))
testthat::expect_error(EML4_sequence(detection,genome="38"))
testthat::expect_error(EML4_sequence(detection,genome=38))

detection <- EML4_ALK_detection(file = file, genome = "hg19")
testthat::expect_true(isa(detection, "GAlignments"))

testthat::expect_error(EML4_sequence(EML4_ALK_detection(file=file),
                                     genome = "hg19"))


testthat::expect_error(EML4_sequence(file))
testthat::expect_error(ALK_sequence(file))
testthat::expect_error(break_position(file))
testthat::expect_error(break_position_depth(reads = file))
testthat::expect_error(EML4_ALK_analysis(file, genome = 38))
testthat::expect_error(EML4_ALK_analysis(file, mates = "38"))
testthat::expect_error(EML4_ALK_analysis(file, genome = "37"))
testthat::expect_error(EML4_ALK_analysis(file, basepairs = "20"))


file <- system.file("extdata","HCC827_EML4.bam", package = "DNAfusion")
introns <- introns_ALK_EML4(file = file)
variants <- find_variants(file = file)

testthat::expect_true(is.character(file))
testthat::expect_true(is.data.frame(introns) | is.character(introns))
testthat::expect_true(is.data.frame(variants) | is.character(variants))

testthat::expect_error(introns_ALK_EML4(file = file, genome = 38))
testthat::expect_error(introns_ALK_EML4(file = file, genome = "hg37"))
testthat::expect_error(find_variants(file = file, genome = 38))
testthat::expect_error(find_variants(file = file, genome = "hg37"))

file <- system.file("extdata","H3122_EML4.bam", package = "DNAfusion")
introns <- introns_ALK_EML4(file = file)
variants <- find_variants(file = file)

testthat::expect_true(is.character(file))
testthat::expect_true(is.data.frame(introns) | is.character(introns))
testthat::expect_true(is.data.frame(variants) | is.character(variants))


R/ALKfusiondiscover_functions.R
#' Detection of EML4-ALK variants
#'
#' This function looks for EML4-ALK mate pair reads in the BAM file.
#'
#' @param file Is the name of the BAM file.
#' @param genome Is the reference genome. Can be either hg38 or hg19. Default = "hg38".
#' @param mates Is the minimum number EML4-ALK mate pairs needed to be detected in order to call a variant. Default = 2.
#' @return If EML4-ALK is detected a data.frame with soft-clipped reads representing EML4-ALK is returned. Otherwise "No EML4-ALK was detected" is returned
#' @examples
#' sample1 <- EML4_ALK_detection(file = "sample1.bam", genome = "hg38", mates = 2)
#' sample2 <- EML4_ALK_detection(file = "sample2.bam", genome = "hg38", mates = 2)
#' @export
EML4_ALK_detection <- function(file, genome = "hg38", mates = 2){
  `%ni%` <- Negate(`%in%`)
  if (genome %ni% c("hg38", "hg19")){
    return("ERROR: The reference genome has to be hg38 or hg19")
  }
  what <- c("mpos", "pos", "seq","cigar")
  if (genome =="hg38"){
    which <- GenomicRanges::GRanges(seqnames="chr2", IRanges::IRanges(start = 42169353, end = 42332548))
    param <- Rsamtools::ScanBamParam(which = which, what = what)
    bam <- Rsamtools::scanBam(file = file, param = param)
    reads <- data.frame(sequences = bam$`chr2:42169353-42332548`$seq,
                        mate = bam$`chr2:42169353-42332548`$mpos,
                        position = bam$`chr2:42169353-42332548`$pos,
                        cigar = bam$`chr2:42169353-42332548`$cigar)
    reads <- reads %>% dplyr::filter(mate < 29921586 & mate > 29192774)
  }
  else{
    which <- GenomicRanges::GRanges(seqnames="chr2", IRanges::IRanges(start = 42396490, end = 42559688))
    param <- Rsamtools::ScanBamParam(which = which, what = what)
    bam <- Rsamtools::scanBam(file = file, param = param)
    reads <- data.frame(sequences = bam$`chr2:42396490-42559688`$seq,
                        mate = bam$`chr2:42396490-42559688`$mpos,
                        position = bam$`chr2:42396490-42559688`$pos,
                        cigar = bam$`chr2:42396490-42559688`$cigar)
    reads <- reads %>% dplyr::filter(mate < 30144477 & mate > 29415640)
  }
  if (length(reads$mate)<mates){
    res <- "No EML4-ALK was detected"
    return(res)
  }
  clip_reads <- reads %>% dplyr::filter(cigar != "96M")
  clip_reads <- clip_reads %>% dplyr::filter(!grepl("D",cigar))
  clip_reads <- clip_reads %>% dplyr::filter(!grepl("I",cigar))
  if (length(clip_reads$mate)<mates){
    res <- "No EML4-ALK was detected"
    return(res)
  }
  return(clip_reads)
}

#' Identification of EML4 breakpoint bases
#'
#' This function identifies the basepairs leading up to the EML4 breakpoint.
#'
#' @param reads Data.frame returned by EML4_ALK_detection().
#' @param basepairs Number of basepairs identified from the EML4-ALK fusion. Default = 20.
#' @return If EML4-ALK is detected, returns a table of identified EML4 basepairs with the number of corresponding reads for each sequence. Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' sample1 <- EML4_sequence(EML4_ALK_detection(file = "sample1.bam", genome = "hg38", mates = 2), basepairs = 20)
#' sample2 <- EML4_sequence(EML4_ALK_detection(file = "sample2.bam", genome = "hg38", mates = 2), basepairs = 20)
#' @export
EML4_sequence <- function(reads, basepairs = 20){
  if(class(reads) != "data.frame"){
    if(reads == "No EML4-ALK was detected"){
      return("No EML4-ALK was detected")
    }
    else{
      return("reads must be a data.frame")
    }
  }
  fun <- function(str) sub("\\M.*", "",str)
  index1 <- sapply(reads$cigar,FUN = fun)
  fun1 <- function(ind1){
    if (length(ind1)>1){
      return(NA)
    }
    else{
      return(ind1)
    }
  }
  fun <- function(ind){
    splits <- strsplit(ind,split = "S")
    splits <- lapply(splits, as.numeric)
    splits_ind <- lapply(splits,FUN = fun1)
    return(splits_ind)
  }
  index2 <- sapply(index1,FUN = fun)
  reads$indeces <- index2
  reads <- reads %>% dplyr::filter(!is.na(indeces))
  EML4_fun <- function(inp){
    return(substring(inp$sequences,(inp$indeces-(basepairs-1)),inp$indeces))
  }
  char_df <-reads %>% dplyr::select(sequences,indeces)
  EML4_seq <- apply(char_df, FUN = EML4_fun, MARGIN = 1)
  EML4_tab <- table(EML4_seq)
  return(EML4_tab)
}

#' Identification of ALK breakpoint bases
#'
#' This function identifies the basepairs following the ALK breakpoint.
#'
#' @param reads Data.frame returned by EML4_ALK_detection().
#' @param basepairs Number of basepairs identified from the EML4-ALK fusion. Default = 20.
#' @return If EML4-ALK is detected, returns a table of identified ALK basepairs with the number of corresponding reads for each sequence. Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' sample1 <- ALK_sequence(EML4_ALK_detection(file = "sample1.bam", genome = "hg38", mates = 2), basepairs = 20)
#' sample2 <- ALK_sequence(EML4_ALK_detection(file = "sample2.bam", genome = "hg38", mates = 2), basepairs = 20)
#' @export
ALK_sequence <- function(reads, basepairs = 20){
  if(class(reads) != "data.frame"){
    if(reads == "No EML4-ALK was detected"){
      return("No EML4-ALK was detected")
    }
    else{
      return("reads must be a data.frame")
    }
  }
  fun <- function(str) sub("\\M.*", "",str)
  index1 <- sapply(reads$cigar,FUN = fun)
  fun1 <- function(ind1){
    if (length(ind1)>1){
      return(NA)
    }
    else{
      return(ind1)
    }
  }
  fun <- function(ind){
    splits <- strsplit(ind,split = "S")
    splits <- lapply(splits, as.numeric)
    splits_ind <- lapply(splits,FUN = fun1)
    return(splits_ind)
  }
  index2 <- sapply(index1,FUN = fun)
  reads$indeces <- index2
  reads <- reads %>% dplyr::filter(!is.na(indeces))
  ALK_fun <- function(inp){
    return(substring(inp$sequences,(inp$indeces+1),(inp$indeces+basepairs)))
  }
  char_df <- reads %>% dplyr::select(sequences,indeces)
  ALK_seq <- apply(char_df, FUN = ALK_fun, MARGIN = 1)
  ALK_tab <- table(ALK_seq)
  return(ALK_tab)
}

#' EML4-ALK breakpoint
#'
#' This function identifies the genomic position in EML4 where the breakpoint has happened.
#'
#' @param reads Data.frame returned by EML4_ALK_detection().
#' @return If EML4-ALK is detected, returns a table of genomic positions with the number of corresponding reads for each sequence. Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' sample1 <- break_position(EML4_ALK_detection(file = "sample1.bam", genome = "hg38", mates = 2))
#' sample2 <- break_position(EML4_ALK_detection(file = "sample2.bam", genome = "hg38", mates = 2))
#' @export
break_position <- function(reads){
  if(class(reads) != "data.frame"){
    if(reads == "No EML4-ALK was detected"){
      return("No EML4-ALK was detected")
    }
    else{
      return("reads must be a data.frame")
    }
  }
  fun <- function(str) sub("\\M.*", "",str)
  index1 <- sapply(reads$cigar,FUN = fun)
  fun1 <- function(ind1){
    if (length(ind1)>1){
      return(NA)
    }
    else{
      return(ind1)
    }
  }
  fun <- function(ind){
    splits <- strsplit(ind,split = "S")
    splits <- lapply(splits, as.numeric)
    splits_ind <- lapply(splits,FUN = fun1)
    return(splits_ind)
  }
  index2 <- sapply(index1,FUN = fun)
  reads$indeces <- index2
  reads <- reads %>% dplyr::filter(!is.na(indeces))
  reads$indeces <- as.numeric(reads$indeces)
  break_pos <- reads$position + reads$indeces-1
  break_pos_tab <- table(break_pos)
  return(break_pos_tab)
}


#' Read depth at breakpoint
#'
#' This function identifies the read depth at the basepair before the breakpoint in EML4.
#'
#' @param file Is the name of the BAM file.
#' @param reads Data.frame returned by EML4_ALK_detection().
#' @return If EML4-ALK is detected a single integer corresponding to the read depth at the breakpoint is returned. Otherwise "No EML4-ALK was detected" is returned
#' @examples
#' sample1 <- break_position_depth(file = "sample1.bam",EML4_ALK_detection(file = "sample1.bam", genome = "hg38", mates = 2))
#' sample2 <- break_position_depth(file = "sample2.bam",EML4_ALK_detection(file = "sample2.bam", genome = "hg38", mates = 2))
#' @export
break_position_depth <- function(file, reads){
  if(class(reads) != "data.frame"){
    if(reads == "No EML4-ALK was detected"){
      return("No EML4-ALK was detected")
    }
    else{
      return("reads must be a data.frame")
    }
  }
  fun <- function(str) sub("\\M.*", "",str)
  index1 <- sapply(reads$cigar,FUN = fun)
  fun1 <- function(ind1){
    if (length(ind1)>1){
      return(NA)
    }
    else{
      return(ind1)
    }
  }
  fun <- function(ind){
    splits <- strsplit(ind,split = "S")
    splits <- lapply(splits, as.numeric)
    splits_ind <- lapply(splits,FUN = fun1)
    return(splits_ind)
  }
  index2 <- sapply(index1,FUN = fun)
  reads$indeces <- index2
  reads <- reads %>% dplyr::filter(!is.na(indeces))
  reads$indeces <- as.numeric(reads$indeces)
  break_pos <- reads$position + reads$indeces-1
  break_pos_tab <- table(break_pos)
  stop_pos <- as.numeric(names(which.max(break_pos_tab)))
  depth <- bamsignals::bamCoverage(file,
                                   GenomicRanges::GRanges(seqnames = "chr2",IRanges::IRanges(start=(stop_pos),end=stop_pos+1)),
                                   mapqual=0,verbose=F)
  return(max(depth[1]))
}

#' Complete EML4-ALK analysis
#'
#' This functions collects the results from the other functions of the package.
#'
#' @param file Is the name of the BAM file.
#' @param genome Is the reference genome. Can be either hg38 or hg19. Default = "hg38".
#' @param mates Is the minimum number EML4-ALK mate pairs needed to be detected in order to call a variant. Default = 2.
#' @param basepairs Number of basepairs identified from the EML4-ALK fusion. Default = 20.
#' @return A list object with clipped_reads corresponding to EML_ALK_detection(), last_EML4 corresponding to EML4_sequence(), first_ALK corresponding to ALK_sequence(), breakpoint corresponding to break_position(), and read_depth corresponding to break_position_depth().
#' @examples
#' sample1 <- EML4_ALK_analysis(file = "sample1.bam", genome = "hg38", mates = 2, basepairs = 20)
#' sample2 <- EML4_ALK_analysis(file = "sample2.bam", genome = "hg38", mates = 2, basepairs = 20)
#' @export
EML4_ALK_analysis <- function(file, genome = "hg38", mates = 2, basepairs = 20){
  res <- EML4_ALK_detection(file = file, genome = genome, mates = mates)
  if(class(res) != "data.frame"){
    return(res)
  }
  EML4 <- EML4_sequence(reads = res, basepairs = basepairs)
  ALK <- ALK_sequence(reads = res, basepairs = basepairs)
  position <- break_position(reads = res)
  position_depth <- break_position_depth(file = file, reads = res)
  return(list(clipped_reads = res,
              last_EML4 = EML4,
              first_ALK = ALK,
              breakpoint = position,
              read_depth = position_depth))
}



R/Genefusiondiscover_functions.R
#' @keywords internal
index_helper <- function(input){
    fun <- function(str){
        result <- sub("\\M.*", "", str)
        splits <- strsplit(result, split = "S")
        if(length(splits[[1]]) > 1){
            splits <- NA
        }
        else{
            splits <- splits
        }
        splits <- lapply(splits, as.numeric)
        return(splits)
    }
    index2 <- vapply(input$cigar, FUN = fun, FUN.VALUE = list(1))
    input$indeces <- index2
    input <- input[!is.na(as.numeric(input[,5])),]
    return(input)
}
#' Detection of EML4-ALK variants
#' 
#' This function looks for EML4-ALK mate pair reads in the BAM file.
#' @import dplyr
#' @importFrom GenomicRanges GRanges
#' @importFrom Rsamtools ScanBamParam scanBam
#' @importFrom IRanges IRanges
#' @param file The name of the file which the data are to be read from.
#' @param genome `Character string` representing the reference genome. 
#' Can be either "hg38" or "hg19". Default="hg38".
#' @param mates `Interger`, the minimum number EML4-ALK mate pairs
#' needed to be detected in order to call a variant. Default=2.
#' @return If EML4-ALK is detected a `data.frame` with soft-clipped reads
#' representing EML4-ALK is returned. 
#' Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' EML4_ALK_detection(file=H3122_bam,
#'                     genome="hg38",
#'                     mates=2)
#' EML4_ALK_detection(file=HCC827_bam,
#'                     genome="hg38",
#'                     mates=2)
#' @export
EML4_ALK_detection <- function(file, genome="hg38", mates=2){
    if(!isa(genome, "character")){
        stop("genome has to be a character")
    }
    if(!isa(mates, "numeric")){
        stop("mates has to be a numeric")
    }
    if(!(genome %in% c("hg38", "hg19"))){
        stop("The reference genome has to be hg38 or hg19")
    }
    what <- c("mpos", "pos", "seq","cigar")
    if (genome =="hg38"){
        which <- GRanges(seqnames="chr2", 
                            IRanges(start=42169353, end=42332548))
        param <- ScanBamParam(which=which, what=what)
        bam <- scanBam(file=file, param=param)
        reads <- as.data.frame(bam[['chr2:42169353-42332548']])
        reads <- reads[(29192774 < reads[,3] & reads[,3] < 29921586 & 
                            !is.na(reads[,3])),]
    }
    else{
        which <- GRanges(seqnames="chr2",
                            IRanges(start=42396490, end=42559688))
        param <- ScanBamParam(which=which, what=what)
        bam <- scanBam(file=file, param=param)
        reads <- as.data.frame(bam[['chr2:42396490-42559688']])
        reads <- reads[(29415640 < reads[,3] & reads[,3] < 30144477 & 
                            !is.na(reads[,3])),]
    }
    if (length(reads[,4])<mates){
        res <- "No EML4-ALK was detected"
        return(res)
    }
    clip_reads <- reads[reads[,2] != "96M",]
    clip_reads <- clip_reads[!grepl("D", clip_reads[,2]),]
    clip_reads <- clip_reads[!grepl("I", clip_reads[,2]),]
    if (length(clip_reads[,4])<mates){
        res <- "No EML4-ALK was detected"
        return(res)
    }
    return(clip_reads)
}

#' Identification of EML4 breakpoint bases
#'
#' This function identifies the basepairs leading up to the EML4 breakpoint.
#'
#' @import dplyr
#' @param reads `Data.frame` returned by EML4_ALK_detection().
#' @param basepairs `Integer`, number of basepairs identified 
#' from the EML4-ALK fusion. Default=20.
#' @return If EML4-ALK is detected, returns a table of identified
#' EML4 basepairs with the number of corresponding reads for each sequence.
#' Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' EML4_sequence(EML4_ALK_detection(file=H3122_bam,
#'                                     genome="hg38",
#'                                     mates=2),
#'                 basepairs=20)
#' EML4_sequence(EML4_ALK_detection(file=HCC827_bam,
#'                                     genome="hg38",
#'                                     mates=2),
#'                 basepairs=20)
#' @export
EML4_sequence <- function(reads, basepairs=20){
    if(!isa(reads, "data.frame")){
        if(reads == "No EML4-ALK was detected"){
            return("No EML4-ALK was detected")
        }
        else{
            stop("reads must be a data.frame")
        }
    }
    if(!isa(basepairs, "numeric")){
        stop("basepairs has to be a numeric")
    }
    reads <- index_helper(reads)
    EML4_fun <- function(inp){
        return(substring(inp[4], (as.numeric(inp[5])-(basepairs-1)),
                            as.numeric(inp[5])))
    }
    EML4_seq <- apply(reads, FUN=EML4_fun, MARGIN=1)
    EML4_tab <- table(EML4_seq)
    return(EML4_tab)
}

#' Identification of ALK breakpoint bases
#'
#' This function identifies the basepairs following the ALK breakpoint.
#'
#' @import dplyr
#' @param reads `data.frame` returned by EML4_ALK_detection().
#' @param basepairs `integer`, number of basepairs identified
#' from the EML4-ALK fusion. Default=20.
#' @return If EML4-ALK is detected, returns a `table` of identified
#' ALK basepairs with the number of corresponding reads for each sequence.
#' Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' ALK_sequence(EML4_ALK_detection(file=H3122_bam,
#'                                 genome="hg38",
#'                                 mates=2),
#'                 basepairs=20)
#' ALK_sequence(EML4_ALK_detection(file=HCC827_bam,
#'                                 genome="hg38",
#'                                 mates=2),
#'                 basepairs=20)
#' @export
ALK_sequence <- function(reads, basepairs=20){
    if(!isa(reads, "data.frame")){
        if(reads == "No EML4-ALK was detected"){
            return("No EML4-ALK was detected")
        }
        else{
            stop("reads must be a data.frame")
        }
    }
    if(!isa(basepairs, "numeric")){
        stop("basepairs has to be a numeric")
    }
    reads <- index_helper(reads)
    ALK_fun <- function(inp){
        return(substring(inp[4], (as.numeric(inp[5])+1),
                            (as.numeric(inp[5])+basepairs)))
    }
    ALK_seq <- apply(reads, FUN=ALK_fun, MARGIN=1)
    ALK_tab <- table(ALK_seq)
    return(ALK_tab)
}

#' EML4-ALK breakpoint
#'
#' This function identifies the genomic position in EML4 
#' where the breakpoint has happened.
#'
#' @import dplyr
#' @param reads Data.frame returned by EML4_ALK_detection().
#' @return If EML4-ALK is detected, returns a `table` of genomic positions
#' with the number of corresponding reads for each sequence. 
#' Otherwise "No EML4-ALK was detected" is returned.
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' break_position(EML4_ALK_detection(file=H3122_bam,
#'                                     genome="hg38",
#'                                     mates=2))
#' break_position(EML4_ALK_detection(file=HCC827_bam,
#'                                     genome="hg38",
#'                                     mates=2))
#' @export
break_position <- function(reads){
    if(!isa(reads, "data.frame")){
        if(reads == "No EML4-ALK was detected"){
            return("No EML4-ALK was detected")
        }
        else{
            stop("reads must be a data.frame")
        }
    }
    reads <- index_helper(reads)
    reads[,5] <- as.numeric(reads[,5])
    break_pos <- reads[,1] + (reads[,5]-1)
    break_pos_tab <- table(break_pos)
    return(break_pos_tab)    
}


#' Read depth at breakpoint
#'
#' This function identifies the read depth at the basepair
#' before the breakpoint in EML4.
#'
#' @import dplyr
#' @importFrom bamsignals bamCoverage
#' @importFrom IRanges IRanges
#' @importFrom GenomicRanges GRanges
#' @param file The name of the file which the data are to be read from.
#' @param reads `data.frame` returned by EML4_ALK_detection().
#' @return If EML4-ALK is detected a single integer corresponding
#' to the read depth at the breakpoint is returned. 
#' Otherwise "No EML4-ALK was detected" is returned
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' break_position_depth(file=H3122_bam,
#'                         EML4_ALK_detection(file=H3122_bam,
#'                                             genome="hg38",
#'                                             mates=2))
#' break_position_depth(file=HCC827_bam,
#'                         EML4_ALK_detection(file=HCC827_bam,
#'                                             genome="hg38",
#'                                             mates=2))
#' @export
break_position_depth <- function(file, reads){
    if(!isa(reads, "data.frame")){
        if(reads == "No EML4-ALK was detected"){
            return("No EML4-ALK was detected")
            }
        else{
            stop("reads must be a data.frame")
            }
    }
    break_pos_tab <- break_position(reads)
    stop_pos <- as.numeric(names(which.max(break_pos_tab)))
    depth <- bamCoverage(file,
                            GRanges(seqnames="chr2",
                            IRanges(start=(stop_pos), end=stop_pos+1)),
                                    mapqual=0, verbose=FALSE)
    return(max(depth[1]))
}

#' Complete EML4-ALK analysis
#'
#' This functions collects the results from the other functions of the package.
#'
#' @import dplyr
#' @param file The name of the file which the data are to be read from.
#' @param genome `character` representing the reference genome. 
#' Can be either "hg38" or "hg19". Default="hg38".
#' @param mates `interger`, the minimum number EML4-ALK mate pairs needed 
#' to be detected in order to call a variant. Default=2.
#' @param basepairs `integer`, number of basepairs identified 
#' from the EML4-ALK fusion. Default=20.
#' @return A `list` object with 
#' clipped_reads corresponding to `EML4_ALK_detection()`, 
#' last_EML4 corresponding to `EML4_sequence()`, 
#' first_ALK corresponding to `ALK_sequence()`, 
#' breakpoint corresponding to `break_position()`, 
#' and read_depth corresponding to `break_position_depth()`.
#' @examples
#' H3122_bam <- system.file("extdata",
#' "H3122_EML4.bam",
#' package="Genefusiondiscover")
#' HCC827_bam <-  system.file("extdata",
#' "HCC827_EML4.bam",
#' package="Genefusiondiscover")
#'
#' EML4_ALK_analysis(file=H3122_bam,
#'                     genome="hg38",
#'                     mates=2,
#'                     basepairs=20)
#' EML4_ALK_analysis(file=HCC827_bam,
#'                     genome="hg38",
#'                     mates=2,
#'                     basepairs=20)
#' @export
EML4_ALK_analysis <- function(file, genome="hg38", mates=2, basepairs=20){
    if(!isa(genome, "character")){
        return("ERROR: genome has to be a character")
    }
    if(!isa(mates, "numeric")){
        return("ERROR: mates has to be a numeric")
    }
    if (!(genome %in% c("hg38", "hg19"))){
        return("ERROR: The reference genome has to be hg38 or hg19")
    }
    res <- EML4_ALK_detection(file=file, genome=genome, mates=mates)
    if(!isa(res, "data.frame")){
        return(res)
    }
    if(!isa(basepairs, "numeric")){
        return("ERROR: basepairs has to be a numeric")
    }
    EML4 <- EML4_sequence(reads=res, basepairs=basepairs)
    ALK <- ALK_sequence(reads=res, basepairs=basepairs)
    position <- break_position(reads=res)
    position_depth <- break_position_depth(file=file, reads=res)
    return(list(clipped_reads=res,
                last_EML4=EML4,
                first_ALK=ALK,
                breakpoint=position,
                read_depth=position_depth))
}

